macros/src/endian_write.rs

```diff
@@ -8,13 +8,20 @@ use syn::{
 
 fn create_get_size_field(field: &Field) -> proc_macro2::TokenStream {
     let field_ident = field.ident.as_ref().expect("Field should have identity");
-    let pad_before = match MacroArgs::from_attributes(&field.attrs) {
-        Some(MacroArgs { pad_before }) => pad_before,
-        _ => 0,
+    let (pad_before, backtrace) = match MacroArgs::from_attributes(&field.attrs) {
+        Some(MacroArgs {
+            pad_before,
+            backtrace,
+        }) => (
+            pad_before.unwrap_or_default(),
+            backtrace.unwrap_or_default(),
+        ),
+        _ => (0, 0),
     };
 
     quote! {
       size += #pad_before;
+      size -= #backtrace;
```

Comment:
`backtrace` tries to solve splitting bytes into bits by rewinding and using the same data twice.  This allows the user to manually separate the data on read and reconcile changes on write.

I like the general idea of having bit level control, although this implementation will result in some amount of data loss if the user doesn't manually reconcile on write.  Currently, users can read and write structs without thinking about how `no_std_io` works, but anything using `backtrace` couldn't do that.

If a solution does get into `no_std_io`, I'd like to make sure it could:
1. Read a struct from a single byte with two fields, each four bits in size
2. Zero out one of the fields
3. Max out the other field (erroring at runtime or compile time if the value is over 15)
4. Write the struct back without issue

Since this is not a common case, I lean towards solving this with a custom struct that manually implements the read/write traits.  This would allow users to enforce their own constraints.

Thoughts?

Reply:
I'm a little lost with the need for manually reconciling on write. If the struct is read and written completely LE or BE then the idea of moving the cursor alone and compensating for the move in the size should be enough for the macro. Otherwise a user would have to manually define the traits anyway. 

Do you have an example of possible data loss?

Reply:
This is the example I was thinking about:
```rs
use no_std_io::Reader;

#[derive(PartialEq, Debug, no_std_io::EndianRead, no_std_io::EndianWrite)]
struct BacktraceTest {
    first: u8,
    #[no_std_io(backtrace = 1)]
    second: u8,
}

#[test]
fn should_not_lose_data() {
    let mut bytes: [u8; 1] = [0xaa];

    let mut backtrace: BacktraceTest = bytes.read_le(0).unwrap();

    // The byte is duplicated between two fields
    assert_eq!(
        backtrace,
        BacktraceTest {
            first: 0xaa,
            second: 0xaa
        }
    );

    // Modify one of the fields
    backtrace.first = 0x00;

    // Write the data back
    bytes.write_le(0, &backtrace);

    // The modified data was not written - the changed data was lost
    assert_eq!(bytes, [0xaa]);
}
```

Reply:
Ah ok, I see what you mean. I find the example a little odd though since in this case it would be two difference fields pointing to the same exact value. As mentioned the user shouldn't have to understand the inner workings of `no_std_io` but if they understand what the attribute does (moves the cursor), then at that point it would be user error.

Reply:
The read test added in this PR has the same issue though - both fields point to the same value, but the first includes a little extra data.  If the first field is zeroed out and the struct is written, there will be data loss with part of the first field's value.

Normally we should add guardrails to prevent user error, but this interface inherently guarantees writing will result in data loss, so there aren't any guardrails we could add.

I have a counter proposal:
Based on the struct in the read test, it looks like there's a `u32` that is intended to use 3 bytes, and one `u8` that follows.

An immediate solution would be to create a `u24` struct that manually implements reading/writing 3 bytes.  It can also guarantee the inner value never exceeds 24 bits in size.

A longer term solution could be a `wrapper` or `from` macro argument:
```rs
#[derive(Debug, PartialEq, EndianRead, EndianWrite)]
struct BacktraceTest {
    #[no_std_io(from = "u24")]
    first: u32,
    second: u8,
}
```

`EndianRead` would read the `from` struct (`u24` in this case), then use `read_value.try_into()?` on the result to convert it to a `u32`.  `EndianWrite` would convert the `u32` to `u24` before writing.  Data loss wouldn't occur, because `try_into`/`try_from` would fail if the value was too large.

This would solve a lot more than the alignment issue:
```rs
#[derive(Debug, PartialEq, EndianRead, EndianWrite)]
struct SomeNexStruct {
    #[no_std_io(from = "NexList<u32>")]
    first: Vec<u32>,
}
```
Here, all of our nex wrappers could be abstracted away.

Reply:
The only thing that is questionable is the definition if what might be considered a new "raw type" that's unaligned. 

I think the use of the attribute should be for moving to where the data is (opposite if pad_before). 

In the case of defining a custom struct, should no_std_io add them or should it be up to the user? With adding something like that, all the normal traits would also have to be applied as the user can't do it themselves.

Also everything between u32 and u64 would have to be added and between u64 and u128.

Reply:
Misaligned data is often a sign of an aligned sub-struct or packed enum, which should be pulled out on their own.  Truly misaligned data shouldn't be common; compilers won't fit a u32 into 24 bits without something extra going on.

Edge cases do exist, like pk1's experience points (possibly the scenario you're looking to solve?), but I want to be careful about adding solutions only used in a small handful of cases.  Every new feature has the potential to make usage, future features, and maintenance harder.

Internally the logic uses a forward moving cursor, but `pad_before` is effectively an alias for adding a hidden `pad: [u8; N]` field, not a cursor interface.  From that perspective, the opposite of `pad_before` is really something like `omit_field`.

Externally, the read/write traits should only be thought of as interfaces to serialize and deserialize data, no matter the implementation.  A macro argument that leaks the implementation details is going to be harder to use since the user needs to be familiar with the implementation.

For example, if someone attempted to use `backtrace` with something similar to `NexList`, it could result in an overflow.  If someone used it next to a pointer type, it would guarantee pointer corruption.  If someone used it next to an enum with 5 variants, the data won't be a valid variant.

We could say these are user errors, but we should try to avoid adding things that can result in a user error, especially if avoiding errors requires implementation knowledge.

I'm not comfortable with `backtrace`, at least as it currently exists, but I do want to help find a solution.

Let's take a step back for a moment to help me understand your need a bit more:
- What is the exact problem you're trying to solve with `backtrace`?
- How come you're looking for something like `backtrace` instead of using a custom struct?

Reply:
Funny enough the reason is because I stumbled upon the exp issue for Gen1 pk files.

My reason for using backtrace is only because defining a "u24" struct just doesn't fit right in my head. My method, if I were to manually read and write, is to move the cursor back and read from the location.

Reply:
Maybe I'm missing something.

The data is a 3 byte number followed by a 1 byte number.  Reading 3 bytes followed by 1 byte matches the data exactly as used by the game, and will always result in correct data.  Writing 3 bytes followed by 1 byte will always write correct data.

Reading 4 bytes, rewinding, then reading 1 byte will result in the first number always being incorrect without post processing.  Writing 4 bytes, rewinding, then writing 1 byte will always result in data loss without preprocessing.  In the worst of cases, these can produce some serious bugs, and implementation knowledge is required to avoid those bugs.

I'm not quite following how:
1. The first scenario doesn't make sense when the data matches perfectly without extra processing, and the second scenario will always read/write incorrect data without extra processing
2. The second scenario is more desirable than the first when every read/write results in reading too much data and overwriting/losing data, and the first scenario always reads and writes the correct amount of data

Would you be able to explain the thoughts behind those?

Reply:
There are many different ways to accomplish the idea here. Whether it's by using functionality of no_std_io or having it be implemented in the library using it. 

What I believe you're seeing is the idea of a custom struct, let's call it `u24`. This struct would contain a `[u8; 3]` that could be used and then converted to an actual type such as `u32`.

Another way of doing this is by reading the four bytes into one field, and then having a get and set method of the underlying seconds.

After actually typing this and thinking about it, you're right. The issue would be once modified. The field would be read as 4 bytes to begin with and then when written to begin with it would be inaccurate not even once it get's overwritten when the cursor of the stream moves.

With this in mind, I believe it may have to be one of the two options mentioned above handled on the side of the library. The idea of backtrace was interesting but might just have to be scrapped. The only other reason I see it existing and when the struct isn't defined in order (but if you are backtracing to the location then why not just put it in order). Even then at that point some sort of order attribute would be better.

Reply:
I think that's a good summary!